File: react-native/react-native-flipper/android/src/main/java/com/facebook/flipper/reactnative/FlipperModule.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.flipper.reactnative;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;

/**
 * The FlipperModule is a React Native Native Module. The instance handles incoming calls that
 * arrive over the React Native bridge from JavaScript. The instance is created per
 * ReactApplicationContext. Which means this module gets reinstated if RN performs a reload. So it
 * should not hold any further state on its own. All state is hold by the Plugin and PluginManager
 * classes.
 */
@ReactModule(name = FlipperModule.NAME)
public class FlipperModule extends ReactContextBaseJavaModule {

  public static final String NAME = "Flipper";

  private FlipperReactNativeJavaScriptPluginManager manager;

  public FlipperModule(
      FlipperReactNativeJavaScriptPluginManager manager, ReactApplicationContext reactContext) {
    super(reactContext);
    this.manager = manager;
  }

  @Override
  public String getName() {
    return NAME;
  }

  @ReactMethod
  public void registerPlugin(
      final String pluginId, final Boolean inBackground, final Callback statusCallback) {
    this.manager.registerPlugin(this, pluginId, inBackground, statusCallback);
  }

  @ReactMethod
  public void send(String pluginId, String method, String data) {
    this.manager.send(pluginId, method, data);
  }

  @ReactMethod
  public void reportErrorWithMetadata(String pluginId, String reason, String stackTrace) {
    this.manager.reportErrorWithMetadata(pluginId, reason, stackTrace);
  }

  @ReactMethod
  public void reportError(String pluginId, String error) {
    this.manager.reportError(pluginId, error);
  }

  @ReactMethod
  public void subscribe(String pluginId, String method) {
    this.manager.subscribe(this, pluginId, method);
  }

  @ReactMethod
  public void respondSuccess(String responderId, String data) {
    this.manager.respondSuccess(responderId, data);
  }

  @ReactMethod
  public void respondError(String responderId, String data) {
    this.manager.respondError(responderId, data);
  }

  public void sendJSEvent(String eventName, WritableMap params) {
    ReactApplicationContext context = getReactApplicationContextIfActiveOrWarn();
    if (context != null) {
      context
          .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
          .emit(eventName, params);
    }
  }
}

File: react-native/react-native-flipper/ios/FlipperModule.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FlipperModule.h"

#import "FlipperReactNativeJavaScriptPluginManager.h"

@implementation FlipperModule {
  __weak FlipperReactNativeJavaScriptPluginManager* _manager;
}

- (instancetype)init {
  return [self initWithManager:[FlipperReactNativeJavaScriptPluginManager
                                   sharedInstance]];
}

- (instancetype)initWithManager:
    (FlipperReactNativeJavaScriptPluginManager*)manager {
  if (self = [super init]) {
    _manager = manager;
  }
  return self;
}

RCT_EXPORT_MODULE(Flipper)

+ (BOOL)requiresMainQueueSetup {
  return NO;
}

- (NSArray<NSString*>*)supportedEvents {
  return @[
    @"react-native-flipper-plugin-connect",
    @"react-native-flipper-plugin-disconnect",
    @"react-native-flipper-receive-event",
  ];
}

- (void)startObserving {
}

- (void)stopObserving {
}

RCT_EXPORT_METHOD(registerPlugin
                  : (NSString*)pluginId inBackground
                  : (BOOL)inBackground statusCallback
                  : (RCTResponseSenderBlock)statusCallback) {
  [_manager registerPluginWithModule:self
                            pluginId:pluginId
                        inBackground:inBackground
                      statusCallback:statusCallback];
}

RCT_EXPORT_METHOD(send
                  : (NSString*)pluginId method
                  : (NSString*)method data
                  : (NSString*)data) {
  [_manager sendWithPluginId:pluginId method:method data:data];
}

RCT_EXPORT_METHOD(reportErrorWithMetadata
                  : (NSString*)pluginId reason
                  : (NSString*)reason stackTrace
                  : (NSString*)stackTrace) {
  [_manager reportErrorWithMetadata:reason
                         stackTrace:stackTrace
                           pluginId:pluginId];
}

RCT_EXPORT_METHOD(reportError : (NSString*)pluginId error : (NSString*)error) {
  [_manager reportError:error pluginId:pluginId];
}

RCT_EXPORT_METHOD(subscribe : (NSString*)pluginId method : (NSString*)method) {
  [_manager subscribeWithModule:self pluginId:pluginId method:method];
}

RCT_EXPORT_METHOD(respondSuccess
                  : (NSString*)responderId data
                  : (NSString*)data) {
  [_manager respondSuccessWithResponderId:responderId data:data];
}

RCT_EXPORT_METHOD(respondError
                  : (NSString*)responderId data
                  : (NSString*)data) {
  [_manager respondErrorWithResponderId:responderId data:data];
}

@end
